command-expansion-server/src/app.ts

```diff
@@ -68,8 +75,38 @@ app.get("/expansion/:expansionId(\\d+)", async (req, res) => {
   return;
 });
 
-app.put("/expansion-set", async (req, res) => {
-  res.status(501).send("PUT /expansion-set: Not implemented");
+app.post("/put-expansion-set", async (req, res) => {
+  const { commandDictionaryId, missionModelId, expansionIds } = req.body.input;
+
+  if (!expansionConfigValidator(req.body.input)) {
+    res.status(400).json({
+      message: `PUT /expansion-set: Invalid request body: ${JSON.stringify(req.body.input)}\n${util.formatWithOptions(
+        { depth: Infinity },
+        expansionConfigValidator.errors
+      )}`,
+    });
+    return;
+  }
+
+  const { rows } = await db.query(`
+    WITH expansion_set_id AS (
+      INSERT INTO expansion_set (command_dict_id, mission_model_id)
+      VALUES (${commandDictionaryId}, ${missionModelId})
+      RETURNING id
+    )
+    INSERT INTO expansion_set_to_rule (set_id, rule_id) VALUES
+      ${expansionIds
+        .map((expansionId: number) => `((SELECT id FROM expansion_set_id), ${expansionId})`)
+        .join(",\n      ")}
+    RETURNING (SELECT id FROM expansion_set_id);
+  `);
+
+  if (rows.length < 1) {
+    throw new Error(`PUT /expansion-set: No expansion set was inserted in the database`);
+  }
+  const id = rows[0].id;
+  console.log(`PUT /expansion-set: Updated expansion set in the database: id=${id}`);
+  res.status(200).json({ expansionSetID: id });
   return;
 });
 
```

Comment:
Should we ignore the rate limiting CodeQL failure? I'm not sure how to accomplish rate limiting with PG or if we want to?

Reply:
@patkenneally and I discussed it and decided that if we do want to implement rate limiting, it would be on exposed APIS (hasura) not this internal one

Reply:
Thanks, in that case I'll dismiss this CodeQL failure